lib/temporal/activity/context.rb

```diff
@@ -84,6 +107,28 @@ def heartbeat(details = nil)
         self
       end
 
+      # Like heartbeat, but a subclass of ActivityInterrupted error will be raised if the
+      # activity has been canceled, start-to-close timeout exceeded, or if the worker is
+      # shutting down. This flag defaults to false, in which these states can be detected by
+      # inspecting the heartbeat response for cancelation, or calling timed_out? or shutting_down?
+      # methods.
+      def heartbeat_interrupted(details = nil)
+        if deadline_exceeded?(schedule_to_close_deadline)
```

Comment:
Semantically it's a bit weird to couple timeouts and shutdowns to the heartbeat mechanism — they don't really have anything to do with one another. Therefore a cancellation mechanism (similar to other SDKs) with a specific exception type might be a better solution overall (see my other comment)

Reply:
They are disparate, but this is basically the single point where activity execution can be pre-empted without resorting to something like `Thread.raise` which is much more disruptive to the existing activity programming model. There is precedent for this in the Java SDK, including guidance on the heartbeat method to rescue/catch the super class of all these errors/exceptions as the proper way to handle interruption.

It sounds like the cancellation mechanism you talk about would allow for one way to be made aware of cancellation/interruption based on any of the 3 pre-emption reasons: worker shutdown, activity cancellation, activity task timeout.